### PR TITLE
Add skip-preview-namelist to case.submit and case_run

### LINKED ISSUE #1276: Add option to skip preview_namelist during case.run
Add an option to case.submit to skip calling preview_namelist during case.run.   This is for low-powered compute nodes that are spending a lot of time in preview_namelist.  Its up to the user to not modify any xml files while the job is in the queue.

### PR DESCRIPTION
**Pull request by erichlf**
Added '--skip-preview-namelist' to case.build

Added '--skip-preview-namelist' to case.build. This will allow the user
to specify if the namelist should be rebuilt from user_nl_*.

Test suite: scripts_regression_tests.py
Test baseline:
Test namelist changes:
Test status: bit for bit

Fixes #1276, ACME-Climate/ACME#1466

User interface changes?: Added --skip-preview-namelist to case.build and
template.case.run

Code review: @jgfouca, @rljacob

### CODE REVIEW

File: scripts/Tools/case.submit [hunk 1]
```diff
@@ -11,26 +11,16 @@ from CIME.case        import Case
...
-%s --test
-
-\033[1mEXAMPLES:\033[0m
-    \033[1;32m# Setup case \033[0m
-    > %s
-""" % ((os.path.basename(args[0]), ) * 4),
-        description=description,
-        formatter_class=argparse.ArgumentDefaultsHelpFormatter
-    )
+    parser = argparse.ArgumentParser(description=description)
```
Comment (jedwards4b) [inline scripts/Tools/case.submit]: Why was this code removed?

File: scripts/Tools/case.submit [hunk 2]
```diff
@@ -45,27 +35,32 @@ OR
...
                         help="Used with tests only, to continue rather than restart a test. ")
 
+    parser.add_argument("--skip-preview-namelist", action="store_true",
+                        help="Skip calling preview-namelist during case.run")
+
     parser.add_argument("-a", "--batch-args",
                         help="Used to pass additional arguments to batch system. ")
 
-    args = parser.parse_args(args[1:])
+    args = parser.parse_args()
```
Comment (jedwards4b) [inline scripts/Tools/case.submit]: This is in conflict with master - I recommend that you rebase and update the PR.

File: scripts/Tools/case.submit [hunk 3]
```diff
@@ -45,27 +35,32 @@ OR
...
 
     CIME.utils.expect(args.prereq is None, "--prereq not currently supported")
 
     CIME.utils.handle_standard_logging_options(args)
 
-    return args.caseroot, args.job, args.no_batch, args.resubmit, args.batch_args
+    return args
```
Comment (jedwards4b) [inline scripts/Tools/case.submit]: We prefer the old method of passing the individual fields of args not the args object itself.

File: scripts/lib/CIME/XML/env_batch.py [hunk 4]
```diff
@@ -43,7 +43,7 @@ def set_value(self, item, value, subgroup=None, ignore_type=False):
                 elif value.count(":") == 2:
                     t_spec = "%H:%M:%S"
                 else:
-                    expect(False, "could not interpret format for wallclock time %s"%value)
+                    expect(False, "could not interpret format for wallclock time {}".format(value))
```
Comment (jedwards4b) [inline scripts/lib/CIME/XML/env_batch.py]: I don't understand why these changes from %s to format were made?

Comment (erichlf) [inline scripts/lib/CIME/XML/env_batch.py]: It is for python3 compatibility. Yes, we aren't using python3, but it is an easy change that doesn't hurt things and will make any future switch to python3 that much easier.

File: scripts/Tools/case.submit (hunk 3 above)
Comment (billsacks) [inline scripts/Tools/case.submit]: "We" is overly broad here. Some of us prefer passing the args object itself :-)

File: scripts/lib/CIME/XML/env_batch.py (hunk 4 above)
Comment (billsacks) [inline scripts/lib/CIME/XML/env_batch.py]: I personally prefer the new-style `format`, because I find it more readable. That said, I'm pretty sure that both are supported by python3.

File: scripts/Tools/case.submit (hunk 3 above)
Comment (jedwards4b) [inline scripts/Tools/case.submit]: I am asking for some uniformity across scripts and we have chosen to pass individual fields rather than the args object.  Same issue with format - we have standardized around %s and I see no reason to change it here.

Comment (billsacks) [inline scripts/Tools/case.submit]: @jedwards4b I see your point about uniformity. I think what's going on here (if I can try to speak somewhat for @erichlf and somewhat for myself) is that there are certain coding standards like these that were put in place when the list of cime developers was very small (basically you and Jim F), and now, as there get to be more cime developers, we're not all in agreement about these coding standards.

Could I suggest, @erichlf - for now leave this as it was, but either open an issue for discussion or put this on the agenda of an upcoming Wednesday cime telecon to revisit our coding standards on these points (`%` vs. format and returning an args object vs. a big list of components).

File: scripts/lib/CIME/XML/env_batch.py (hunk 4 above)
Comment (erichlf) [inline scripts/lib/CIME/XML/env_batch.py]: I think you are right that both are currently supported, but if I recall the plan is not to support the old style.

Comment (jedwards4b) [inline scripts/lib/CIME/XML/env_batch.py]: I do not object to changes that bring us closer to python3 compatibility - however I do object to having a mixture of %s and format and we are about to do a release.  I would prefer to wait until after the release and then apply this change uniformly throughout the code.

**Comment by erichlf** [comment]
I believe I have addressed all issues, and now things should be working with no conflicts.

File: scripts/lib/CIME/XML/env_batch.py (hunk 4 above)
Comment (erichlf) [inline scripts/lib/CIME/XML/env_batch.py]: `{}` and `format` have already been extensively implemented throughout the code and not by me.

File: scripts/lib/CIME/case_run.py [hunk 5]
```diff
@@ -120,7 +121,7 @@ def _run_model_impl(case, lid):
                         restore_from_archive(case)
 
                         orig_cont = case.get_value("CONTINUE_RUN")
-                        if not orig_cont:
+                        if not (orig_cont or skip_pnl):
```
Comment (jgfouca) [inline scripts/lib/CIME/case_run.py]: Maybe @jedwards4b can help me here. I don't know if skipping create_namelists here is safe if orig_cont is False. run_model can be invoked several times for certain tests and in other situations.

Comment (jedwards4b) [inline scripts/lib/CIME/case_run.py]: You still want to change the CONTINUE_RUN value in either case, so the skip_pnl should be a seperate conditional
if not orig_cont:
   case.set_value("CONTINUE_RUN", True)
   if not skp_pnl:
       create_namelists(case)

File: scripts/lib/CIME/XML/env_batch.py (hunk 4 above)
Comment (erichlf) [inline scripts/lib/CIME/XML/env_batch.py]: Absolutely happy to do so.

File: config/acme/machines/template.case.run [hunk 6]
```diff
@@ -32,21 +32,21 @@ import argparse, doctest
...
 OR
-%s --test
+{} --test
 
 \033[1mEXAMPLES:\033[0m
     \033[1;32m# case.run SMS\033[0m
-    > %s
-""" % ((os.path.basename(args[0]), ) * 4),
+    > {}
+""".format(((os.path.basename(args[0]), ) * 4),
```
Comment (erichlf) [inline config/acme/machines/template.case.run]: Will fix. In this case it would be `{0}` since lists/arrays start at 0 in python.

File: config/acme/machines/template.case.run [hunk 7]
```diff
@@ -56,11 +56,16 @@ formatter_class=argparse.ArgumentDefaultsHelpFormatter
...
     args = CIME.utils.parse_args_and_handle_standard_logging_options(args, parser)
+    parser.add_argument("--skip-preview-namelist", action="store_true",
+                        help="Skip calling preview-namelist during case.run")
 
     if args.caseroot is not None:
         os.chdir(args.caseroot)
 
-    return args.caseroot
+    if args.skip_preview_namelist is None:
+      args.skip_preview_namelist = False
```
Comment (erichlf) [inline config/acme/machines/template.case.run]: doh!

File: scripts/lib/CIME/case_run.py (hunk 5 above)
Comment (erichlf) [inline scripts/lib/CIME/case_run.py]: Will fix.

File: scripts/lib/CIME/case_run.py [hunk 8]
```diff
@@ -122,6 +123,7 @@ def _run_model_impl(case, lid):
                         orig_cont = case.get_value("CONTINUE_RUN")
                         if not orig_cont:
                             case.set_value("CONTINUE_RUN", True)
+                        if not skip_pnl:
```
Comment (jgfouca) [inline scripts/lib/CIME/case_run.py]: This still isn't right. You've changed the original intent which was to re-run create_namelists if CONTINUE_RUN was False. For now, let's not worry about skip_pnl in this code. This code is for recovery from node failures which is a rare corner case and we don't need to worry about how it performs.

**Comment by erichlf** [comment]
This should be ready now.

**Comment by jgfouca** [comment]
@erichlf why did you assign this to @rljacob ? Usually Jim or I are the integrators.

**Comment by erichlf** [comment]
@jgfouca Because he was on the origin issue.

**Comment by rljacob** [comment]
With the new reviewing system in github, I'm not sure it matters.  As long as you or Jim review it.  But to get it merged quickly, probably want to re-assign it.

**Comment by ndkeen** [comment]
When might we see this in ACME?

**Comment by rljacob** [comment]
We are currently a little past alpha.10 we'd need alpha.12 to get this.
